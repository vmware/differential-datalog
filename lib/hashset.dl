/* Immutable hash sets.
 * This module contains bindings for the `HashSet` type
 * from the `im` crate. */

#[iterate_by_ref=iter:'A]
extern type HashSet<'A>

extern function hashset_singleton(x: 'X): HashSet<'X>
extern function hashset_empty(): HashSet<'X>

function size(s: HashSet<'X>): usize {
    hashset_size(s)
}

function insert(s: mut HashSet<'X>, v: 'X) {
    hashset_insert(s, v)
}

function insert_imm(s: HashSet<'X>, v: 'X): HashSet<'X> {
    hashset_insert_imm(s, v)
}

function remove(s: mut HashSet<'X>, v: 'X) {
    hashset_remove(s, v)
}

function remove_imm(s: HashSet<'X>, v: 'X): HashSet<'X> {
    hashset_remove_imm(s, v)
}

function contains(s: HashSet<'X>, v: 'X): bool {
    hashset_contains(s, v)
}

function is_empty(s: HashSet<'X>): bool {
    hashset_is_empty(s)
}

function nth(s: HashSet<'X>, n: usize): Option<'X> {
    hashset_nth(s, n)
}

function to_vec(s: HashSet<'A>): Vec<'A> {
    hashset_to_vec(s)
}

function to_hashset(v: Vec<'A>): HashSet<'A> {
    var res = hashset_empty();
    for (x in v) {
        res.insert(x);
    };
    res
}

function to_hashset(g: Group<'K, 'A>): HashSet<'A> {
    var res = hashset_empty();
    for ((x, _) in g) {
        res.insert(x);
    };
    res
}

function to_hashset(o: Option<'X>): HashSet<'X> {
    match (o) {
        Some{x} -> hashset_singleton(x),
        None -> hashset_empty()
    }
}

function union(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X> {
    hashset_union(s1, s2)
}

function union(sets: Vec<HashSet<'X>>): HashSet<'X> {
    hashset_unions(sets)
}

function union(sets: Group<'K, HashSet<'X>>): HashSet<'X> {
    group_hashset_unions(sets)
}

function intersection(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X> {
    hashset_intersection(s1, s2)
}

function difference(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X> {
    hashset_difference(s1, s2)
}

/* Applies closure `f` to each element of the set. */
function map(s: HashSet<'A>, f: function('A): 'B): HashSet<'B> {
    var res = hashset_empty();
    for (x in s) {
        res.insert(f(x))
    };
    res
}

/* Returns the element that gives the minimum value from the specified function.
 * If several elements are equally minimum, the first element is returned.
 * If the set is empty, `None` is returned. */
function arg_min(s: HashSet<'A>, f: function('A): 'B): Option<'A> {
    hashset_arg_min(s, f)
}

/* Returns the element that gives the maximum value from the specified function.
 * If several elements are equally maximum, the first element is returned.
 * If the set is empty, `None` is returned. */
function arg_max(s: HashSet<'A>, f: function('A): 'B): Option<'A> {
    hashset_arg_max(s, f)
}

/* Returns the first element of the set that satisfies predicate `f` or
 * `None` if none of the elements satisfy the predicate. */
function find(s: HashSet<'A>, f: function('A): bool): Option<'A> {
    for (x in s) {
        if (f(x)) {
            return Some{x}
        }
    };
    None
}

/* Returns a set containing only those elements in `s` that satisfy predicate
 * `f`. */
function filter(s: HashSet<'A>, f: function('A): bool): HashSet<'A> {
    var res = hashset_empty();
    for (x in s) {
        if (f(x)) {
            res.insert(x)
        }
    };
    res
}

/* Both filters and maps the set.
 *
 * Calls the closure on each element of the set.  If the closure returns
 * `Some{element}`, then that element is returned. */
function filter_map(s: HashSet<'A>, f: function('A): Option<'B>): HashSet<'B> {
    var res = hashset_empty();
    for (x in s) {
        match (f(x)) {
            None -> (),
            Some{y} -> res.insert(y)
        }
    };
    res
}

/* Returnds `true` iff all elements of the set satisfy predicate `f`. */
function all(s: HashSet<'A>, f: function('A): bool): bool {
    for (x in s) {
        if (not f(x)) {
            return false
        }
    };
    true
}

/* Returnds `true` iff at least one element of the set satisfies predicate `f`. */
function any(s: HashSet<'A>, f: function('A): bool): bool {
    for (x in s) {
        if (f(x)) {
            return true
        }
    };
    false
}

/* Iterates over the set is ascending order, aggregating its contents using `f`.
 *
 * `f` - takes the previous value of the accumulator and the next element in the
 * set and returns the new value of the accumulator.
 *
 * `initializer` - initial value of the accumulator. */
function fold(s: HashSet<'A>, f: function('B, 'A): 'B, initializer: 'B): 'B {
    var res = initializer;
    for (x in s) {
        res = f(res, x)
    };
    res
}

extern function hashset_arg_min(s: HashSet<'A>, f: function('A): 'B): Option<'A>
extern function hashset_arg_max(s: HashSet<'A>, f: function('A): 'B): Option<'A>
extern function hashset_size(s: HashSet<'X>): usize
extern function hashset_insert(s: mut HashSet<'X>, v: 'X)
extern function hashset_remove(s: mut HashSet<'X>, v: 'X)
extern function hashset_insert_imm(s: HashSet<'X>, v: 'X): HashSet<'X>
extern function hashset_remove_imm(s: HashSet<'X>, v: 'X): HashSet<'X>
extern function hashset_contains(s: HashSet<'X>, v: 'X): bool
extern function hashset_is_empty(s: HashSet<'X>): bool
extern function hashset_nth(s: HashSet<'X>, n: usize): Option<'X>
extern function hashset_to_vec(s: HashSet<'A>): Vec<'A>
extern function hashset_union(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X>
extern function hashset_unions(sets: Vec<HashSet<'X>>): HashSet<'X>
extern function group_hashset_unions(sets: Group<'K, HashSet<'X>>): HashSet<'X>
extern function hashset_intersection(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X>
extern function hashset_difference(s1: HashSet<'X>, s2: HashSet<'X>): HashSet<'X>
