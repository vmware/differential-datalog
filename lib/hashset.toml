[dependencies.im]
git = "https://github.com/ddlog-dev/im-rs"
branch = "ddlog-1"
features = ["serde"]
