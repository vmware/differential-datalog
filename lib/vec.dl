/*
 * Functions for working with vectors (see basic vector operations defined in
 * `ddlog_std.dl`).
 */

/* Applies closure `f` to each element of the vector. */
function map(v: Vec<'A>, f: function('A): 'B): Vec<'B> {
    var res = vec_with_capacity(v.len());
    for (x in v) {
        res.push(f(x))
    };
    res
}

/* Applies closure `f` to each element of the vector and concatenates the
 * resulting vectors, returning a flat vector. */
function flatmap(v: Vec<'A>, f: function('A): Vec<'B>): Vec<'B> {
    var res = vec_empty();
    for (x in v) {
        res.append(f(x))
    };
    res
}

/* Sorts the vector with a key extraction function, but may not preserve the
 * order of equal elements. */
function sort_by(v: mut Vec<'A>, f: function('A): 'B) {
    vec_sort_by(v, f)
}

/* Returns the element that gives the minimum value from the specified function.
 * If several elements are equally minimum, the first element is returned.
 * If the vector is empty, `None` is returned. */
function min_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
    vec_min_by(v, f)
}

/* Returns the element that gives the maximum value from the specified function.
 * If several elements are equally maximum, the first element is returned.
 * If the vector is empty, `None` is returned. */
function max_by(v: Vec<'A>, f: function('A): 'B): Option<'A> {
    vec_max_by(v, f)
}

/* Returns the first element of the vector that satisfies predicate `f` or
 * `None` if none of the elements satisfy the predicate. */
function find(v: Vec<'A>, f: function('A): bool): Option<'A> {
    for (x in v) {
        if (f(x)) {
            return Some{x}
        }
    };
    None
}

/* Returns a vector containing only those elements in `v` that satisfy predicate `f`,
 * preserving the order of the elements in the original vector. */
function filter(v: Vec<'A>, f: function('A): bool): Vec<'A> {
    var res = vec_empty();
    for (x in v) {
        if (f(x)) {
            res.push(x)
        }
    };
    res
}

/* Both filters and maps the vector.
 *
 * Calls the closure on each element of the vector.  If the closure returns
 * `Some{element}`, then that element is returned. */
function filter_map(v: Vec<'A>, f: function('A): Option<'B>): Vec<'B> {
    var res = vec_empty();
    for (x in v) {
        match (f(x)) {
            None -> (),
            Some{y} -> res.push(y)
        }
    };
    res
}


/* Retains only the elements specified by the predicate.
 * In other words, remove all elements `e` such that `f(e)` returns `false`.
 * This method operates in place, visiting each element exactly once in the
 * original order, and preserves the order of the retained elements. */
function retain(v: mut Vec<'A>, f: function('A): bool) {
    var del: s64 = 0;
    var len = v.len();

    for (i in range(0, (len as isize) - 1, 1)) {
        var x = v.nth(i as usize).unwrap_or_default();
        if (not f(x)) {
            del = del + 1
        } else if (del > 0) {
            v.update_nth((i - del) as usize, x);
        }
    };
    if (del > 0) {
        v.truncate(len - (del as usize));
    }
}

/* Returns `true` iff all elements of the vector satisfy predicate `f`. */
function all(v: Vec<'A>, f: function('A): bool): bool {
    for (x in v) {
        if (not f(x)) {
            return false
        }
    };
    true
}

/* Returns `true` iff at least one element of the vector satisfies predicate `f`. */
function any(v: Vec<'A>, f: function('A): bool): bool {
    for (x in v) {
        if (f(x)) {
            return true
        }
    };
    false
}

/* Counts the number of elements in `v` that satisfy predicate `f`. */
function count(v: Vec<'A>, f: function('A): bool): usize {
    var cnt = 0;
    for (x in v) {
        if (f(x)) {
            cnt = cnt + 1
        }
    };
    cnt
}

/* Iterates over the vector, aggregating its contents using `f`.
 *
 * `f` - takes the previous value of the accumulator and the next element in the
 * vector and returns the new value of the accumulator.
 *
 * `initializer` - initial value of the accumulator. */
function fold(v: Vec<'A>, f: function('B, 'A): 'B, initializer: 'B): 'B {
    var res = initializer;
    for (x in v) {
        res = f(res, x)
    };
    res
}

extern function vec_sort_by(v: mut Vec<'A>, f: function('A): 'B)
extern function vec_min_by(v: Vec<'A>, f: function('A): 'B): Option<'A>
extern function vec_max_by(v: Vec<'A>, f: function('A): 'B): Option<'A>
