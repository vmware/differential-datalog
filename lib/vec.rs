use crate::closure::Closure;
use crate::ddlog_std;

pub fn vec_sort_by<A, B: Ord>(v: &mut ddlog_std::Vec<A>, f: &Box<dyn Closure<*const A, B>>) {
    v.x.sort_unstable_by_key(|x| f.call(x))
}

pub fn vec_min_by<A: Clone, B: Ord>(
    v: &ddlog_std::Vec<A>,
    f: &Box<dyn Closure<*const A, B>>,
) -> ddlog_std::Option<A> {
    ddlog_std::Option::from(v.x.iter().min_by_key(|x| f.call(*x)).map(|x| x.clone()))
}

pub fn vec_max_by<A: Clone, B: Ord>(
    v: &ddlog_std::Vec<A>,
    f: &Box<dyn Closure<*const A, B>>,
) -> ddlog_std::Option<A> {
    ddlog_std::Option::from(v.x.iter().max_by_key(|x| f.call(*x)).map(|x| x.clone()))
}
