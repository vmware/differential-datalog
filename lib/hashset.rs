/*
Copyright (c) 2021 VMware, Inc.
SPDX-License-Identifier: MIT

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

use ddlog_rt::Closure;
use ddlog_std::{option2std, tuple2, Group, Option as DDOption, Vec as DDVec};

use differential_datalog::record::{CollectionKind, Record};
use im::hashset::{ConsumingIter, HashSet as IMHashSet, OrderedConsumingIter, OrderedIter};
use serde::{
    de::Deserializer,
    ser::{SerializeSeq, Serializer},
};
use std::{
    borrow::Borrow,
    cmp::Ordering,
    collections::hash_map::DefaultHasher,
    fmt::{Debug, Formatter, Result as FmtResult},
    hash::{BuildHasherDefault, Hash, Hasher},
    iter::FromIterator,
    ops::DerefMut,
    result::Result as StdResult,
};

#[derive(Clone)]
pub struct HashSet<T> {
    // We must use a deterministic hasher here instead of the default `RandomState`
    // to ensure that hashset iterators yield elements in a consistent order.
    set: IMHashSet<T, BuildHasherDefault<DefaultHasher>>,
}

impl<T> Default for HashSet<T> {
    fn default() -> Self {
        HashSet {
            set: IMHashSet::with_hasher(<BuildHasherDefault<DefaultHasher>>::default()),
        }
    }
}

impl<T: Hash + Eq + Ord> Hash for HashSet<T> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Our modified `im::HashSet` implementation ensures that `hash` is
        // deterministic, i.e., given the same hasher and two sets with the
        // same elements it returns the same hash value.  In contrast, the
        // original `im` crate currently violates this property.
        self.set.hash(state);
    }
}

impl<T: Hash + Eq + Ord> PartialEq for HashSet<T> {
    fn eq(&self, other: &Self) -> bool {
        // The `Eq` implementation in `im::HashSet` is unable to rely on the
        // ordered iterator (which requires the `Ord` trait) and is therefore
        // inefficient compared to this.

        // Check if `other` is a clone of `self`.
        if self.set.ptr_eq(&other.set) {
            return true;
        }
        if self.len() != other.len() {
            return false;
        }
        self.iter().eq(other.iter())
    }
}

impl<T: Hash + Eq + Ord> Eq for HashSet<T> {}

impl<T: Hash + Eq + Clone + PartialOrd + Ord> PartialOrd for HashSet<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        if self.set.ptr_eq(&other.set) {
            return Some(Ordering::Equal);
        }
        // We don't need to implement lexicographical ordering.  If the two
        // sets have different sizes, just compare the sizes.
        match self.len().cmp(&other.len()) {
            Ordering::Equal => (),
            ord => return Some(ord),
        }
        // Our modified `im::HashSet` implementation ensures that `partial_cmp`
        // is consistent: given the same hasher and two sets with the
        // same elements, it returns `Ordering::Equal`.  In contrast, the
        // original `im` crate currently violates this property.
        self.set.partial_cmp(&other.set)
    }
}

impl<T: Hash + Eq + Clone + Ord> Ord for HashSet<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        if self.set.ptr_eq(&other.set) {
            return Ordering::Equal;
        }
        match self.len().cmp(&other.len()) {
            Ordering::Equal => (),
            ord => return ord,
        }

        // Our modified `IMHashSet` implementation ensures that `cmp`
        // is consistent: given the same hasher and two sets with the
        // same elements it returns `Ordering::Equal`.  In contrast, the
        // original `im` crate currently violates this property.
        self.set.cmp(&other.set)
    }
}

impl<T: Serialize + Hash + Eq + Clone + Ord> Serialize for HashSet<T> {
    fn serialize<S>(&self, serializer: S) -> StdResult<S::Ok, S::Error>
    where
        S: Serializer,
    {
        // Don't use IMHashSet's implementation of `serialize` to
        // ensure deterministic order.
        let mut s = serializer.serialize_seq(Some(self.len()))?;
        for i in self.iter() {
            s.serialize_element(i)?;
        }
        s.end()
    }
}

impl<'de, T: Deserialize<'de> + Hash + Eq + Clone> Deserialize<'de> for HashSet<T> {
    fn deserialize<D>(deserializer: D) -> StdResult<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        IMHashSet::deserialize(deserializer).map(HashSet::from)
    }
}

impl<T> HashSet<T> {
    pub fn len(&self) -> usize {
        self.set.len()
    }
}

impl<'a, T: Hash + Ord> HashSet<T> {
    // Always use this iterator over `im::HashSet::iter` in scenarios where
    // order matters (serialization, formatting, etc.).
    pub fn iter(&'a self) -> OrderedIter<'a, T> {
        // The default IMHashSet iterator does not guarantee consistent
        // ordering: two sets with the same elements can iterate over them
        // in different order even when using the same hasher.  This can
        // lead to non-deterministic behavior in DDlog code that iterates
        // over hashsets.  We must therefore use a deterministic iterator.
        self.set.ordered_iter()
    }
}

impl<T: Hash + Eq + Clone + Ord> IntoIterator for HashSet<T> {
    type Item = T;
    type IntoIter = OrderedConsumingIter<T>;

    fn into_iter(self) -> Self::IntoIter {
        // Use ordered iterator that guarantees consistent ordering.
        self.set.into_ordered_iter()
    }
}

impl<T: Hash + Eq + Clone + From<RT>, RT> FromIterator<RT> for HashSet<T> {
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = RT>,
    {
        HashSet {
            set: IMHashSet::from_iter(iter),
        }
    }
}

impl<T> HashSet<T> {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn is_empty(&self) -> bool {
        self.set.is_empty()
    }
}

impl<T> From<IMHashSet<T, BuildHasherDefault<DefaultHasher>>> for HashSet<T> {
    fn from(set: IMHashSet<T, BuildHasherDefault<DefaultHasher>>) -> Self {
        HashSet { set }
    }
}

impl<T: Hash + Eq> HashSet<T> {
    pub fn contains<BT: ?Sized>(&self, x: &BT) -> bool
    where
        BT: Hash + Eq,
        T: Borrow<BT>,
    {
        self.set.contains(x)
    }
}

impl<T: Hash + Eq + Clone> HashSet<T> {
    pub fn unit(x: T) -> Self {
        let mut set = IMHashSet::default();
        set.insert(x);
        HashSet::from(set)
    }

    pub fn insert(&mut self, x: T) -> Option<T> {
        self.set.insert(x)
    }

    pub fn remove<BT: ?Sized>(&mut self, x: &BT) -> Option<T>
    where
        BT: Hash + Eq,
        T: Borrow<BT>,
    {
        self.set.remove(x)
    }

    pub fn without(&self, x: &T) -> Self {
        HashSet::from(self.set.without(x))
    }

    pub fn update(&self, x: T) -> Self {
        HashSet::from(self.set.update(x))
    }

    pub fn union(self, other: Self) -> Self {
        // `im::HashSet` implements union by adding elements from `other` to
        // `self`.  This is inefficient when `other` is much
        // larger than `self`.  The following optimization will become
        // unnecessary once this PR or something similar has landed:
        // https://github.com/bodil/im-rs/pull/163.
        if self.len() >= other.len() {
            HashSet::from(self.set.union(other.set))
        } else {
            HashSet::from(other.set.union(self.set))
        }
    }
    pub fn unions<I>(i: I) -> Self
    where
        I: IntoIterator<Item = Self>,
    {
        let mut sets: Vec<_> = i.into_iter().collect();

        // Start with the largest set.
        if let Some((largest_idx, _)) = sets.iter().enumerate().max_by_key(|(_, s)| s.len()) {
            let mut largest_set = Self::default();
            std::mem::swap(&mut largest_set, &mut sets[largest_idx]);

            sets.into_iter().fold(largest_set, Self::union)
        } else {
            Self::default()
        }
    }
    pub fn intersection(self, other: Self) -> Self {
        HashSet::from(self.set.intersection(other.set))
    }
    pub fn difference(self, other: Self) -> Self {
        HashSet::from(self.set.difference(other.set))
    }
}

// I don't want to implement `Deref` for `HashSet` to prevent
// accidentally calling `im::HashSet` methods.  For example,
// `im::HashSet::iter` returns the non-deterministic iterator (which can
// mess up DDlog in ways one can easily miss unless they test on big
// enough sets).

// impl<T> Deref for HashSet<T> {
//     type Target = IMHashSet<T>;
//
//     fn deref(//self) -> //Self::Target {
//         //self.set
//     }
// }
//
// impl<T> DerefMut for HashSet<T> {
//     fn deref_mut(//mut self) -> //mut Self::Target {
//         //mut self.set
//     }
// }

impl<T: FromRecord + Hash + Eq + Clone + Ord> FromRecordInner for HashSet<T> {
    fn from_record_inner(val: &Record) -> StdResult<Self, String> {
        match val {
            Record::Array(_, args) => StdResult::from_iter(args.iter().map(T::from_record)),
            v => T::from_record(v).map(Self::unit),
        }
    }
}

impl<T: IntoRecord + Hash + Eq + Clone + Ord> IntoRecord for HashSet<T> {
    fn into_record(self) -> Record {
        Record::Array(
            CollectionKind::Set,
            self.into_iter().map(|x| x.into_record()).collect(),
        )
    }
}

// Set update semantics: update contains values that are in one of the sets
// but not the other.
impl<T: FromRecord + ::serde::de::DeserializeOwned + Hash + Eq + Clone + Ord>
    MutatorInner<HashSet<T>> for Record
{
    fn mutate_inner(&self, set: &mut HashSet<T>) -> StdResult<(), String> {
        let upd = <HashSet<T>>::from_record(self)?;
        *set = HashSet::from(set.set.clone().symmetric_difference(upd.set));
        Ok(())
    }
}

impl<T: Hash + Eq + Debug + Ord> Debug for HashSet<T> {
    fn fmt(&self, f: &mut Formatter) -> FmtResult {
        f.debug_set().entries(self.iter()).finish()
    }
}

pub fn hashset_size<T>(s: &HashSet<T>) -> std_usize {
    s.len() as std_usize
}

pub fn hashset_empty<T>() -> HashSet<T> {
    HashSet::new()
}

pub fn hashset_singleton<T: Hash + Eq + Clone>(v: &T) -> HashSet<T> {
    HashSet::unit((*v).clone())
}

pub fn hashset_insert<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
    s.insert(v.clone());
}

pub fn hashset_insert_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
    s.update((*v).clone())
}

pub fn hashset_remove<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
    s.remove(v);
}

pub fn hashset_remove_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
    s.without(v)
}

pub fn hashset_contains<T: Hash + Eq>(s: &HashSet<T>, v: &T) -> bool {
    s.contains(v)
}

pub fn hashset_is_empty<T>(s: &HashSet<T>) -> bool {
    s.is_empty()
}

pub fn hashset_nth<T: Hash + Clone + Ord>(s: &HashSet<T>, n: &std_usize) -> DDOption<T> {
    option2std(s.iter().nth(*n as usize).cloned())
}

pub fn hashset_to_vec<T: Hash + Clone + Ord>(s: &HashSet<T>) -> DDVec<T> {
    DDVec {
        vec: s.iter().cloned().collect(),
    }
}

pub fn hashset_union<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
    s1.clone().union(s2.clone())
}

pub fn hashset_unions<T: Hash + Eq + Clone + Ord>(sets: &Vec<HashSet<T>>) -> HashSet<T> {
    HashSet::unions(sets.iter().cloned())
}

pub fn group_hashset_unions<K, T: Hash + Eq + Clone + Ord>(
    sets: &Group<K, HashSet<T>>,
) -> HashSet<T> {
    HashSet::unions(sets.iter().map(|tuple2(x, _)| x))
}

pub fn hashset_intersection<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
    s1.clone().intersection(s2.clone())
}

pub fn hashset_difference<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
    s1.clone().difference(s2.clone())
}

pub fn hashset_arg_min<T: Hash + Clone + Ord, B: Ord>(
    s: &HashSet<T>,
    f: &Box<dyn Closure<*const T, B>>,
) -> DDOption<T> {
    DDOption::from(s.iter().min_by_key(|x| f.call(*x)).map(|x| (*x).clone()))
}

pub fn hashset_arg_max<T: Hash + Clone + Ord, B: Ord>(
    s: &HashSet<T>,
    f: &Box<dyn Closure<*const T, B>>,
) -> DDOption<T> {
    DDOption::from(s.iter().max_by_key(|x| f.call(*x)).map(|x| (*x).clone()))
}
