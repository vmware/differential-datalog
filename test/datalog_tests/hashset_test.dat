dump hashset_test::SetTransforms;
dump hashset_test::SetFolds;

start;

insert hashset_test::TestSetMutateI(1, [1,2,3,4,5]),
insert hashset_test::TestSetMutateI(2, [1,2,3,4,5]),
insert hashset_test::TestSetMutateI(3, [1,2,3,4,5]),
insert hashset_test::TestSetMutateI(4, [1,2,3,4,5]),
insert hashset_test::TestSetMutateI(5, [1,2,3,4,5]),

commit dump_changes;

start;

modify hashset_test::TestSetMutateI 1 <- hashset_test::TestSetMutateI{1, [1,2,3,4,5]},
modify hashset_test::TestSetMutateI 2 <- hashset_test::TestSetMutateI{2, [6,7,8,9,10]},
modify hashset_test::TestSetMutateI 3 <- hashset_test::TestSetMutateI{3, [1,2,3]},
modify hashset_test::TestSetMutateI 4 <- hashset_test::TestSetMutateI{4, [1,3,5,7,9]},
modify hashset_test::TestSetMutateI 5 <- hashset_test::TestSetMutateI{5, []},

commit dump_changes;
