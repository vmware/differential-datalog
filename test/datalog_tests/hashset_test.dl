import hashset
import vec

typedef TestSet = HashSet<(s64, Vec<s64>)>

function test_set(): TestSet {
    [ (0, [0,1,2])
    , (1, [2,3,4])
    , (2, [4,5,6])
    , (3, [6,7,8])
    , (4, [8,9,10])
    ].to_hashset()
}


output relation SetTransforms(description: string, s: TestSet)

// map.
SetTransforms(
    "map(push 100)",
    test_set().map(|x| { var y = x; y.1.push(100); y } )).

// filter.
SetTransforms(
    "filter(contains 2)",
    test_set().filter(|x| x.1.any(|y| y == 2))).

// filter_map.
SetTransforms(
    "filter_map(==1, :=100)",
    test_set().filter_map(|x| if (x.0 == 1) {
        Some{ {var y = x; y.0 = 100; y} }
    } else {
        None
    })).

// arg_min.
SetTransforms(
    "arg_min(min)",
    hashset_singleton(test_set().arg_min(|x| x.1.arg_min(|x|x)).unwrap_or_default())).

// arg_max.
SetTransforms(
    "arg_max(max)",
    hashset_singleton(test_set().arg_max(|x| x.1.arg_max(|x|x)).unwrap_or_default())).

// find.
SetTransforms(
    "find(contains(3))",
    test_set().find(|x| x.1.any(|y|y==3)).to_hashset()).

// any.
SetTransforms(
    "any(contains(3))",
    if (test_set().any(|x| x.1.any(|y|y==3))) {
        test_set()
    } else {
        hashset_empty()
    }).

// all.
SetTransforms(
    "all(contains(3))",
    if (test_set().all(|x| x.1.any(|y|y==3))) {
        test_set()
    } else {
        hashset_empty()
    }).

// fold
output relation SetFolds(description: string, a: s64)

SetFolds(
    "fold(+)",
    test_set().fold(|a, x| a + x.0, 0)
).

SetFolds(
    "fold(fold(+))",
    test_set().fold(|a, x| a + x.1.fold(|b,y| b + y, 0), 0)
).

input relation TestSetMutateI(k: usize, v: HashSet<usize>)
primary key (x) x.k

output relation TestSetMutateO(k: usize, v: HashSet<usize>)

TestSetMutateO(k, v) :- TestSetMutateI(k, v).
