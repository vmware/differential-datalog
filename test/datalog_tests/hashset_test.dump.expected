hashset_test::SetTransforms{.description = "all(contains(3))", .s = []}
hashset_test::SetTransforms{.description = "any(contains(3))", .s = [(2, [4, 5, 6]), (4, [8, 9, 10]), (3, [6, 7, 8]), (1, [2, 3, 4]), (0, [0, 1, 2])]}
hashset_test::SetTransforms{.description = "arg_max(max)", .s = [(4, [8, 9, 10])]}
hashset_test::SetTransforms{.description = "arg_min(min)", .s = [(0, [0, 1, 2])]}
hashset_test::SetTransforms{.description = "filter(contains 2)", .s = [(1, [2, 3, 4]), (0, [0, 1, 2])]}
hashset_test::SetTransforms{.description = "filter_map(==1, :=100)", .s = [(100, [2, 3, 4])]}
hashset_test::SetTransforms{.description = "find(contains(3))", .s = [(1, [2, 3, 4])]}
hashset_test::SetTransforms{.description = "map(push 100)", .s = [(1, [2, 3, 4, 100]), (3, [6, 7, 8, 100]), (2, [4, 5, 6, 100]), (0, [0, 1, 2, 100]), (4, [8, 9, 10, 100])]}
hashset_test::SetFolds{.description = "fold(+)", .a = 10}
hashset_test::SetFolds{.description = "fold(fold(+))", .a = 75}
hashset_test::TestSetMutateO:
hashset_test::TestSetMutateO{.k = 1, .v = [4, 2, 1, 3, 5]}: +1
hashset_test::TestSetMutateO{.k = 2, .v = [4, 2, 1, 3, 5]}: +1
hashset_test::TestSetMutateO{.k = 3, .v = [4, 2, 1, 3, 5]}: +1
hashset_test::TestSetMutateO{.k = 4, .v = [4, 2, 1, 3, 5]}: +1
hashset_test::TestSetMutateO{.k = 5, .v = [4, 2, 1, 3, 5]}: +1
hashset_test::TestSetMutateO:
hashset_test::TestSetMutateO{.k = 1, .v = []}: +1
hashset_test::TestSetMutateO{.k = 1, .v = [4, 2, 1, 3, 5]}: -1
hashset_test::TestSetMutateO{.k = 2, .v = [4, 2, 1, 3, 5]}: -1
hashset_test::TestSetMutateO{.k = 2, .v = [10, 6, 8, 9, 4, 2, 1, 3, 7, 5]}: +1
hashset_test::TestSetMutateO{.k = 3, .v = [4, 5]}: +1
hashset_test::TestSetMutateO{.k = 3, .v = [4, 2, 1, 3, 5]}: -1
hashset_test::TestSetMutateO{.k = 4, .v = [9, 4, 2, 7]}: +1
hashset_test::TestSetMutateO{.k = 4, .v = [4, 2, 1, 3, 5]}: -1
